tools/opam-updates.py

```diff
@@ -0,0 +1,326 @@
+
+    return list(filtered_out)
+
+
+# Argument parsing
+
+class AuthAction(argparse.Action):
+    """
+    Argparse type action for auth command line arguments of the form:
+       USER:PASSWORD
+    Returns a (USER, PASSWORD) tuple if successful, otherwise raises
+    ArgumentTypeError.
+    """
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        auth = tuple(values.split(':', 1))
+
+        if len(auth) != 2:
+            msg = "malformed auth passed to --auth: %r" % values
+            raise argparse.ArgumentTypeError(msg)
+
+        setattr(namespace, self.dest, auth)
+
+
+def get_netrc_auth():
+    "Try to get the right auth from netrc."
+    try:
+        login, _, password = netrc().authenticators("github.com")
+        return [(login, password)]
+    except (IOError, ValueError, NetrcParseError):
+        return []
+
+
+def parse_args_or_exit(argv=None):
+    """
+    Parse command line options.
+    """
+    parser = argparse.ArgumentParser(
+        description="xs-opam package upgrade tool")
+
+    parser.add_argument("xs_opam", metavar="XS-OPAM", nargs="?", default=None,
+                        help="Path to the local clone of the xs-opam repository. "
+                        "If omitted, the script will do a dry-run.")
+
+    # In principle we don't need a default as requests can use netrc on
+    # its own, however we need to know if we can find an out or now, so
+    # we infer it by ourselves.
+    parser.add_argument("--auth", metavar="AUTH",
+                        default=get_netrc_auth(), action=AuthAction,
+                        help="GitHub API authentication in the form "
+                        "USER:TOKEN, where TOKEN is a Personal Access Token "
+                        "(see https://github.com/settings/tokens). "
+                        "If omitted, the script will try to use the "
+                        "$HOME/.netrc file.")
+
+    parser.add_argument("--ignore-xs", dest="ignore_xs", action="store_true",
+                        help="Do not check if we are trying to "
+                        "upgrade packages in `xs` or `xs-extra`.")
+
+    args = parser.parse_args(argv)
+
+    if args.auth is None and args.xs_opam is not None:
+        sys.exit("Authentication is needed to be able to use the github api")
+
+    return args
+
+
+# Main body
+
+def main(argv=None):
+    """
+    Entry point.
+    """
+    args = parse_args_or_exit(argv)
+
+    # compute useful paths
+    if args.xs_opam is not None:
+        upstream_p = os.path.join(args.xs_opam, "packages", "upstream")
+        upstream_extra_p = os.path.join(
+            args.xs_opam, "packages", "upstream-extra")
+        xs_p = os.path.join(args.xs_opam, "packages", "xs")
+        xs_extra_p = os.path.join(args.xs_opam, "packages", "xs-extra")
+
+    out = get_opam_upgrade_output()
+
+    upgrades = [
+        extract_upgrade(row) for row in out
+        if "- upgrade" in row and "xenctrl" not in row
```

Comment:
Hmm, how does this match the arrow symbol:
```
  ↗  upgrade   cppo               1.6.0 to 1.6.1
```

Reply:
When running in a subprocess it only uses ASCII